src/structs/lepton_header.rs

```diff
@@ -95,77 +102,81 @@ impl LeptonHeader {
+        // header[5..8] is reserved
 
-        // We use 12 bytes of git revision for our needs - mark that it's C# implementation and a not-compressed header size.
-        self.uncompressed_lepton_header_size = 0;
-        if header[5] == 'M' as u8 && header[6] == 'S' as u8 {
-            c.set_position(7);
-            self.uncompressed_lepton_header_size = c.read_u32::<LittleEndian>()?;
+        // header[8..20] 12 bytes were the GIT revision, but for historical reasons we
+        // also use this space to store the uncompressed lepton header size plus some
+        // flags to detect the SIMD flavor that was used to encode, since
+        // previously the encoder would generate different incompatible files depending on
+        // whether SIMD or scalar was selected by the build options.
+        if header[8] == 'M' as u8 && header[9] == 'S' as u8 {
+            self.uncompressed_lepton_header_size =
+                Some(u32::from_le_bytes(header[10..14].try_into().unwrap()));
 
             // read the flag bits to know how we should decode this file
-            let flags = c.read_u8()?;
+            let flags = header[14];
             if (flags & 0x80) != 0 {
                 enabled_features.use_16bit_dc_estimate = (flags & 0x01) != 0;
                 enabled_features.use_16bit_adv_predict = (flags & 0x02) != 0;
             }
+
+            self.encoder_version = header[15];
+            self.git_revision_prefix = header[16..20].try_into().unwrap();
+        } else {
+            // take first bytes for git revision prefix
+            self.git_revision_prefix = header[8..12].try_into().unwrap();
         }
 
-        // full size of the original file
-        c.set_position(17);
-        self.plain_text_size = c.read_u32::<LittleEndian>()?;
+        // total size of original JPEG
+        self.jpeg_file_size = u32::from_le_bytes(header[20..24].try_into().unwrap());
 
-        // now read the compressed header
-        let compressed_header_size = reader.read_u32::<LittleEndian>()? as usize;
+        let compressed_header_size =
+            u32::from_le_bytes(header[24..28].try_into().unwrap()) as usize;
 
+        Ok(compressed_header_size)
+    }
+
+    /// reads the start of the lepton file and parses the compressed header. Returns the raw JPEG header contents.
+    pub fn read_compressed_lepton_header<R: Read>(
+        &mut self,
+        reader: &mut R,
+        enabled_features: &mut EnabledFeatures,
+        compressed_header_size: usize,
+    ) -> Result<()> {
         if compressed_header_size > MAX_FILE_SIZE_BYTES as usize {
             return err_exit_code(ExitCode::BadLeptonFile, "Too big compressed header");
         }
-        if self.plain_text_size > MAX_FILE_SIZE_BYTES as u32 {
+        if self.jpeg_file_size > MAX_FILE_SIZE_BYTES as u32 {
             return err_exit_code(ExitCode::BadLeptonFile, "Only support images < 128 megs");
         }
```

Comment:
Why keep this limit? Removing this I successfully encoded and decoded
https://upload.wikimedia.org/wikipedia/commons/3/30/The_most_complete_and_accurate_reconstruction_of_Girl_with_a_Pearl_Earring-21_Gigapixel-Digital_Profoundism-Demo.jpg

Reply:
I guess we can add this to enabled_features instead. Next change... this one is getting too big.